build/i2c/src/lib.rs

```diff
@@ -1037,7 +1046,26 @@ impl ConfigGenerator {
             let id = sensors.len();
             sensors.push(kind);
 
-            let name: Option<String> = if let Some(power) = &d.power {
+            // In most cases, when the power member variable is present, sensors
+            // have a one-to-one association with power rails.  However, this
+            // isn't always true: in the power shelf, for example, there are two
+            // rails and three (uncorrelated) temperature sensors.
+            //
+            // This is indicated with the `sensors` array, which allows us to
+            // specify only certain kinds of sensors being tied to rails.
+            //
+            // If the `sensors` array is `None`, then we fall back to the
+            // default case of all sensors being one-to-one associated with
+            // rails.
+            let name: Option<String> = if d.power.is_some()
```

Comment:
I took a bunch of stabs at making this easier to read and mostly failed. One possibility that might work: what if we extract this into a separate function? Something like

```rust
        // In most cases, ... rest of that comment ...
        fn power_for_kind(kind: Sensor, d: &I2cDevice) -> Option<&I2cPower> {
            d.power.as_ref().and_then(|power| {
                if let Some(sensors) = power.sensors.as_ref() {
                    if !sensors.contains(&kind) {
                        return None;
                    }
                }
                Some(power)
            })
        }
```

at which point I think this could be

```rust

let name: Option<String> = if let Some(power) = power_for_kind(kind, d) {
    if let Some(rails) = &power.rails {
        // .. rest unchanged ..
```

Reply:
Yup, brought this into a helper function under `I2cDevice` and cleaned it up _substantially_.